src/qr_login.rs

```diff
@@ -0,0 +1,141 @@
+    pub fn new(
+        public_key: Curve25519PublicKey,
+        rendezvous_url: &str,
+        homeserver_url: Option<String>,
+    ) -> Result<QrCodeData, JsError> {
+        let public_key = public_key.inner;
+        let rendezvous_url = Url::parse(rendezvous_url)?;
+
+        let mode_data = if let Some(homeserver_url) = homeserver_url {
+            qr_login::QrCodeModeData::Reciprocate { homeserver_url: Url::parse(&homeserver_url)? }
+        } else {
+            qr_login::QrCodeModeData::Login
+        };
+
+        let inner = qr_login::QrCodeData { public_key, rendezvous_url, mode_data };
+
+        Ok(QrCodeData { inner })
+    }
+
+    /// Attempt to decode a slice of bytes into a [`QrCodeData`] object.
+    ///
+    /// The slice of bytes would generally be returned by a QR code decoder.
+    pub fn from_bytes(bytes: &[u8]) -> Result<QrCodeData, JsError> {
+        Ok(Self { inner: qr_login::QrCodeData::from_bytes(bytes)? })
+    }
+
+    /// Encode the [`QrCodeData`] into a list of bytes.
+    ///
+    /// The list of bytes can be used by a QR code generator to create an image
+    /// containing a QR code.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.inner.to_bytes()
+    }
+
+    /// Attempt to decode a base64 encoded string into a [`QrCodeData`] object.
+    pub fn from_base64(data: &str) -> Result<QrCodeData, JsError> {
+        Ok(Self { inner: qr_login::QrCodeData::from_base64(data)? })
+    }
+
+    /// Encode the [`QrCodeData`] into a string using base64.
+    ///
+    /// This format can be used for debugging purposes and the
+    /// [`QrcodeData::from_base64()`] method can be used to parse the string
+    /// again.
+    pub fn to_base64(&self) -> String {
+        self.inner.to_base64()
+    }
+
+    /// Get the Curve25519 public key embedded in the [`QrCodeData`].
+    ///
+    /// This Curve25519 public key should be used to establish an
+    /// [ECIES](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme)
+    /// (Elliptic Curve Integrated Encryption Scheme) channel with the other
+    /// device.
+    #[wasm_bindgen(getter)]
+    pub fn public_key(&self) -> Curve25519PublicKey {
+        self.inner.public_key.into()
+    }
+
+    /// Get the URL of the rendezvous server which will be used to exchange
+    /// messages between the two devices.
+    #[wasm_bindgen(getter)]
+    pub fn rendezvous_url(&self) -> String {
+        self.inner.rendezvous_url.as_str().to_owned()
+    }
+
+    /// Get the homeserver URL which the new device will be logged in to.
+    ///
+    /// This will be only available if the existing device has generated the QR
+    /// code and the new device is the one scanning the QR code.
+    #[wasm_bindgen(getter)]
+    pub fn homeserver_url(&self) -> Option<String> {
+        if let qr_login::QrCodeModeData::Reciprocate { homeserver_url } = &self.inner.mode_data {
+            Some(homeserver_url.as_str().to_owned())
+        } else {
+            None
+        }
+    }
```

Comment:
When we're adding new methods to the WASM bindings, please can we use `#[wasm_bindgen(js_name)]` to give them camelCase names, to be consistent with the rest of the bindings?